MultiVault

kumavis opened the pull request:
*status: undergoing final QA*

DO NOT MERGE INTO MASTER

danfinlay committed:
Add multi-vault planning doc

danfinlay committed:
Add to doc

danfinlay committed:
Add initial KeyringController files

danfinlay committed:
Merge branch 'master' into i328-MultiVault

danfinlay committed:
Remove opinionated seed word code

Completely breaking all account maangement, I have removed the opinionated seed-word code from the UI.

Web3 injection still seems to work.

danfinlay committed:
Remove additional deprecated action

danfinlay committed:
Add minimal method signatures to new keyring controller

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into origin/i328-MultiVault

danfinlay committed:
Began adding browser-native encryptor module

Added new Qunit build process that will browserify the contents of `test/integration/lib` into the QUnit browser, allowing much more modular testing, including unit testing of our modules in our target browsers.

Made a basic unit test file of this form for the new encryptor module, which fails miserably because I've only just begun to work with it.

I've started with this blog post as a starting point, and will be adjusting it to our needs from there:
http://qnimate.com/passphrase-based-encryption-using-web-cryptography-api/

danfinlay committed:
Got encrypting working, not yet decrypting

danfinlay committed:
Got basic encryptor working

danfinlay committed:
Fixed vector usage in encryptor

danfinlay committed:
Linted

2-am-zzz committed:
Initial work on UI side

2-am-zzz committed:
Merge in crypto.

2-am-zzz committed:
Add crypto packages to package.json

2-am-zzz committed:
Remove unused crypto dependencies.

2-am-zzz committed:
Intermediary commit.

2-am-zzz committed:
Git merge resolution.

danfinlay committed:
Add debug flag to dev mode

2-am-zzz committed:
Resolve merge conflict.

danfinlay committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

danfinlay committed:
Linted

danfinlay committed:
Change actions reference to keyringController to background

danfinlay committed:
Commented out useless advanced not on init screen

danfinlay committed:
Fix unlock logic

danfinlay committed:
Fix state updating after vault creation and unlocking

danfinlay committed:
Linted & added salting to vault

danfinlay committed:
Fix tests

danfinlay committed:
Fix background connection references

2-am-zzz committed:
Merge branch 'master' into i328-MultiVault

danfinlay committed:
Merge branch 'master' into i328-MultiVault

danfinlay committed:
Began adding first basic keyring

danfinlay committed:
Improved simple account generation

danfinlay committed:
Added tx and msg signing to keychain & controller

danfinlay committed:
Added SimpleKeyring tests

danfinlay committed:
Create basic keyring-controller unit test file

danfinlay committed:
Create new vault before each keyringController unit test

danfinlay committed:
Fix bug in new KeyringController vault restoring logic.

danfinlay committed:
Added tx & msg managing functionality to new KeyringController

danfinlay committed:
Fix loading indication

danfinlay committed:
Fix duplicate provider menu item bug

danfinlay committed:
Merge branch 'master' into i328-MultiVault

2-am-zzz committed:
Merge branch 'master' into i328-MultiVault

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

2-am-zzz committed:
Phase out extra warning screen.

2-am-zzz committed:
Remove debugger statement.

danfinlay committed:
Merge branch 'master' into i328-MultiVault

danfinlay committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

danfinlay committed:
Add first-time state

danfinlay committed:
Add latest jazzicon version

danfinlay committed:
Update first screen to match design.

danfinlay committed:
Fix identicon rendering

danfinlay committed:
Add hd wallet keyring

danfinlay committed:
Added hd wallet tests

danfinlay committed:
Configure BIP44 Keychain as default one

danfinlay committed:
Add vault migration test

2-am-zzz committed:
Merge branch 'master' into i328-MultiVault

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

danfinlay committed:
Now migrating old vaults to new DEN format

2-am-zzz committed:
Modify first-screen design to match docs.

2-am-zzz committed:
Add password warning div.

2-am-zzz committed:
Add focus to password field.

2-am-zzz committed:
Make case condition more consistent. Remove extra action. Fix incorrect display of network.

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

2-am-zzz committed:
Implement auto-fauceting

2-am-zzz committed:
Implement seed word confirmation page.
Remove logs.
Move HD render files to ui/app.

2-am-zzz committed:
Linted

2-am-zzz committed:
Do not open MM page in dev mode.

2-am-zzz committed:
Made progress on parity for MultiVault
- Deleted some unused items
- Renamed files and paths to match with new locations.
- Modified keyring controller logic to separate concerns.
- Fix account naming issues.
- Enable creation of new vault with default HD keyring.
- Formatting issues.

danfinlay committed:
Fix seed phrase restore

danfinlay committed:
Remove and untrack npm debug files

danfinlay committed:
Make forgot password button.

danfinlay committed:
Fixed bugs related to clearing caches when restoring to a new vault

danfinlay committed:
Make identicon tolerant of changing state

2-am-zzz committed:
Add reveal of seed words.

danfinlay committed:
Fix handling of migrating old vault style

Now old vaults are recognized as an "Initialized" MetaMask instance.

Upon logging in, when fetching the initial password-derived key, if there is no new-style vault, but there is an old style vault, it is migrated to the new format before proceeding through the usual unlocking steps.

danfinlay committed:
Fix initial vault creation

danfinlay committed:
Fix dual vault creation bug

danfinlay committed:
Don't faucet in development

danfinlay committed:
Don't faucet during tests

danfinlay committed:
Set test env during tests

danfinlay committed:
Fix vault creation bug

danfinlay committed:
Fix account nicknaming bug

When nicknaming, we weren't normalizing the input, and so we were retrieving with differently formatted addresses than we were persisting.

danfinlay committed:
Finish fixing nicknaming bug

2-am-zzz committed:
Implement private key exporting.

danfinlay committed:
Fix password validation and persistence issue

Was wiping the vault on each successful password attempt... :P

danfinlay committed:
Add tolerance to normalize method

danfinlay committed:
Remove insane actions hash

danfinlay committed:
Fix network checking

danfinlay committed:
Fix 787 gas buffer bug

danfinlay committed:
Fix BN reference

danfinlay committed:
Fix nickname test

danfinlay committed:
Fix integration test suite

danfinlay committed:
Merge branch 'master' into i328-MultiVault

Commit:
Merge branch 'master' into i328-MultiVault

Commit:
fix tx data so it has the network id

2-am-zzz committed:
Merge pull request #802 from MetaMask/confTxFix

Conf tx fix

kumavis committed:
Merge branch 'master' into i328-MultiVault

kumavis committed:
Merge branch 'master' into i328-MultiVault

2-am-zzz committed:
Linting to the max.

2-am-zzz committed:
Typo fix.

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    this.configManager.setSeedWords(seedWords)
+  }
+
+  submitPassword (password, cb) {
```
used in both: unlocking metamask, showing seed words

calls `migrateAndGetKey` - overly close couppling of migration and controller

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    return this.encryptor.keyFromPassword(password + salt)
+    .then((key) => {
+      this.key = key
+      this.configManager.setSalt(salt)
```
salt should be serialized in vault, not config

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+  addNewKeyring (type, opts, cb) {
+    const Keyring = this.getKeyringClassForType(type)
+    const keyring = new Keyring(opts)
+    const accounts = keyring.getAccounts()
```
not sure if accounts is available synchronously? should keyring be an event emitter and emit new keys that we then push to setup

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+
+  addNewAccount (keyRingNum = 0, cb) {
+    const ring = this.keyrings[keyRingNum]
+    const accounts = ring.addAccounts(1)
```
potentially async

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+
+  // Takes an account address and an iterator representing
+  // the current number of named accounts.
+  loadBalanceAndNickname (account) {
```
"loadBalance" is confusing with "loading a balance"

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    this.createNickname(address)
+  }
+
+  createNickname (address) {
```
should this be the role of the keyring?

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    }
+  }
+
+  persistAllKeyrings () {
```
I think persistence and encryption should happen outside this controller

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    ], didComplete)
+
+    // perform static analyis on the target contract code
+    function analyzeForDelegateCall (cb) {
```
ideally move to transaction manager

kumavis reviewed (commented):
we should have 1 or 2 comments for every method explaining what its used for

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+const hdkey = require('ethereumjs-wallet/hdkey')
+const bip39 = require('bip39')
+const ethUtil = require('ethereumjs-util')
+const type = 'HD Key Tree'
```
move to after requires

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+
+const hdPathString = `m/44'/60'/0'/0`
+
+module.exports = class HdKeyring extends EventEmitter {
```
we should have a generic base class

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+
+module.exports = class HdKeyring extends EventEmitter {
+
+  static type () {
```
static properties are done  like this
```js
Point.ZERO = new Point(0, 0)
```

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+    }
+  }
+
+  initFromMnemonic (mnemonic) {
```
organize into standard KeyRing method overrides and type-specific methods

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+    return rawMsgSig
+  }
+
+  getWalletForAccount (account) {
```
what is a `wallet`

kumavis replied:
not sure that this can be part of the keyring standard as we cant extract a private key from remote id mgmt like treznor

kumavis commented on app/scripts/keyrings/hd.js:
```diff
@@ -0,0 +1,104 @@
+  }
+}
+
+function normalize (address) {
```
extract to util fn

kumavis commented on app/scripts/lib/encryptor.js:
```diff
@@ -0,0 +1,149 @@
+  generateSalt,
+}
+
+// Takes a Pojo, returns encrypted text.
```
`cypher text`

kumavis commented on app/scripts/lib/encryptor.js:
```diff
@@ -0,0 +1,149 @@
+var ethUtil = require('ethereumjs-util')
```
* we dont need to do anything ethereum specific
* we shouldnt follow ethereums lead of inefficient encoding of binary as hex (prefer base64)
* this should be a seperate module, in fact we should likely use some off-the-shelf module

kumavis commented on app/scripts/lib/sig-util.js:
```diff
@@ -0,0 +1,23 @@
+const ethUtil = require('ethereumjs-util')
```
what is this file for

kumavis commented on app/scripts/metamask-controller.js:
```diff
@@ -11,15 +11,18 @@ const extension = require('./lib/extension')
     this.ethStore = new EthStore(this.provider)
-    this.idStore.setStore(this.ethStore)
+    this.keyringController.setStore(this.ethStore)
+    this.getNetwork()
```
rename - this sounds like it returns the network

kumavis commented on test/integration/index.js:
```diff
@@ -0,0 +1,21 @@
+var fs = require('fs')
```
sort of recreating our build sys here
the concerns me

kumavis commented on ui/app/components/pending-tx.js:
```diff
@@ -30,22 +30,22 @@ PendingTx.prototype.render = function () {
-          },
-        }, 'Transaction Error. Exception thrown in contract code.')
+      txData.simulationFails
+      ? h('.error', {
```
lets move the `?`s back to the previous line

kumavis commented:
I think we should continue this refactor to clean up large architectural problems.
This will require more weeks of effort and QA but should massively reduce the complexity of the code.

In order to not block hotfixes, we should merge this into 'dev'. However I can't change the target branch of this PR nor can i migrate the review comments to a new PR, so lets keep this PR active and finally merge it into dev.

Here are the phases of this continued work:
* address comments here
* merge into `dev`
* continue architectural refactor

kumavis commented:
architectural refactor points:
* turn stateful things into observable stores
* move persistence outside of controllers
* move ui state transition to background

2-am-zzz reviewed (changes requested):
Blocked for merging for awhile, until changes are addressed, both code-wise and architecture-wise.

kumavis commented:
Pardon my critical nature, im just concerned about the complexity of our code base - i think the work so far is a great step forward, and now we understand what we need, we can make the larger shifts required to have a lean, functional, intuitive codebase

Zanibas commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+      unconfMsgs: messageManager.unconfirmedMsgs(),
+      messages: messageManager.getMsgList(),
+      selectedAddress: address,
+      selectedAccount: address,
```
We need to do a clearer audit of what properties we want to keep on this object.

2-am-zzz committed:
Fix lint rules for ternary operator placement.

2-am-zzz committed:
Package normalize inside sig util file.

2-am-zzz committed:
Clarify functions names. Package normalize inside util file. Fix require headers.

2-am-zzz committed:
Various keyring changes
- Rearrange the require statements.
- Remove unused web3.
- Simplify some callbacks and promises.
- Clarify certain parameters.

2-am-zzz committed:
Revert to previous anonymous function for promise.

2-am-zzz committed:
Rename n variables to numberOfAccounts inside of test suite. Remove idStore tests.

2-am-zzz committed:
Add idStore tests again.

2-am-zzz committed:
Re-enable disclaimer screen. Rename variables to reflect role more clearly.

Zanibas commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+      seedWords: this.configManager.getSeedWords(),
+      isInitialized: (!!wallet || !!vault),
+      isUnlocked: !!this.key,
+      isConfirmed: true, // AUDIT this.configManager.getConfirmed(),
```
Doesn't seem to be hiccuping right now. Tentatively putting it back in until we determine why we set it to true in the first place.

kumavis commented:
@Zanibas nice nice

2-am-zzz committed:
Minimize repeated code.

2-am-zzz committed:
Minimize repeated code.

kumavis commented on app/scripts/keyring-controller.js:
```diff
@@ -172,9 +171,11 @@ module.exports = class KeyringController extends EventEmitter {
     const firstKeyring = this.keyrings[0]
+    console.log(firstKeyring)
     const seedWords = firstKeyring.serialize().mnemonic
+    console.log(seedWords)
```
pls remove console log

Zanibas replied:
whoops~

Zanibas commented:
Whoops

On Tue, Nov 15, 2016, 13:28 kumavis <notifications@github.com> wrote:

> *@kumavis* commented on this pull request.
> ------------------------------
>
> In app/scripts/keyring-controller.js
> <https://github.com/MetaMask/metamask-plugin/pull/806#pullrequestreview-8700378>
> :
>
> >      const seedWords = firstKeyring.serialize().mnemonic
> +    console.log(seedWords)
>
> pls remove console log
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MetaMask/metamask-plugin/pull/806#pullrequestreview-8700378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwTuWKUqrph8XBz7n1Vh37jk9Rgb1uRks5q-iP_gaJpZM4Kvc7H>
> .
>

kumavis commented:
@Zanibas can you solve the merge conflicts

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

2-am-zzz committed:
Resolve merge conflicts.

2-am-zzz committed:
Resolve merge conflicts.

2-am-zzz committed:
Merge branch 'i328-MultiVault' of github.com:MetaMask/metamask-plugin into i328-MultiVault

2-am-zzz committed:
Lints.

2-am-zzz committed:
Fix semantics in comments

Zanibas commented:
Are we able to trim down idStore to the essentials needed for migration?

2-am-zzz committed:
Minimize dispatches by using emitters and relying on state updates.

2-am-zzz committed:
Revert one cb to previous state.

Zanibas commented on README.md:
```diff
@@ -90,6 +90,10 @@ You can also test with a continuously watching process, via `npm run watch`.
 
```
Starting the review!

flyswatter commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+      return cb('Seed phrase is invalid.')
+    }
+
+    this.clearKeyrings()
```
This doesn't clear the vault (persisted, encrypted storage), it clears the unencrypted RAM storage.  We could maybe rename better, and add a comment to help clarify that goal.

flyswatter commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,566 @@
+    const salt = this.encryptor.generateSalt()
+    configManager.setSalt(salt)
+
+    return this.migrateAndGetKey(password)
```
this should just be `this.loadKey(password)` instead since we never need to care about an old vault when creating a new vault.

flyswatter commented on app/scripts/keyring-controller.js:
```diff
@@ -0,0 +1,572 @@
+    .then(() => {
+      return this.persistAllKeyrings()
+    })
+    .then(() => {
```
I'm ok with the adapter lib, but `then(cb)` would throw error to cb.
